### DISCUSSION TITLE
How to unit test with 2.0 style transactions

### DISCUSSION THREAD

**Discussion by mcclurem** [sqlalchemy/sqlalchemy#7752] [Usage Questions] [answered]
A common usage pattern I personally use and I've seen others use for unit testing code is to wrap all of a unit test inside a transaction save point to be able to roll the whole thing back. With the removal of sub transactions this won't be possible. In addition 1.4.x broke portions of this behavior with the context checks on sessions

pre 1.4 example:
```python
@pytest.fixture(autouse=True)
def db(app, monkeypatch):
    '''Wraps the test in a database transaction and rolls it back at teardown.'''
    from myapp.db import db as _db
    connection = _db.engine.connect()
    connection.begin()
    transaction = connection.begin_nested()

    # Patch Flask-SQLAlchemy to use our connection
    monkeypatch.setattr(_db, 'get_engine', lambda *args: connection)

    @event.listens_for(_db.session, "after_transaction_end")
    def restart_savepoint(session, transaction):
        if transaction.nested and not transaction.parent.nested:
            session.expire_all()
            session.begin_nested()

    try:
        yield _db
    finally:
        _db.session.remove()
        transaction.rollback()
        connection.close()
```

Starting with 1.4, this only works if you change the `restart_savepoint` function to perform `connection.begin_nested()` instead of `session.begin_nested()` (otherwise `engine/util.py::TransactionalContext._trans_ctx_check` will bite you). I expect this is even further broken with 2.0 style elimination of nested transactions though I have not tested it further.

I understand that nested transactions created this complicated environment where `session.commit()` wasn't guaranteed to do any persisting. However the /one/ use case where it really seems unavoidable is in unit testing where you genuinely want the ability to intercept commits so that you can eventually revert them at tear down.

Are there any suggestions for alternative methodologies here? Perhaps we need some core sqlalchemy support to codify this paradigm for unit testing specifically?

**Comment by zzzeek** [comment, marked as answer]
we have a complete example of the Session w/ savepoint transaction nesting pattern compatible with 1.4 and 2.0, as well as with 1.3, should work in all versions, seen at https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#joining-a-session-into-an-external-transaction-such-as-for-test-suites .   If not, please report a bug with a complete [MCVE](http://stackoverflow.com/help/mcve) that uses no dependencies and we will make it work. Thanks!

**Comment by mcclurem** [comment]
Ah thanks so much @zzzeek ... sorry I missed the RTFM. Just to clarify: is the reason that this example works because the connection owns the savepoint not the session?

The reason I thought this would be broken is because of the comments in the "Commit as you go" paragraph in: https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#session-level-vs-engine-level-transaction-control

The indication seemed to be that either session.commit or connection.commit would do the same thing and so this only seems to make sense if it said something like: "Session.commit will always commit the outermost transaction for that session, and Connection.commit will always commit the outermost transaction for that connection (which will include the session which is bound)"

**Reply by zzzeek** [reply]
yes the connection owns the savepoint but also the Session itself detects that the transaction came in from the outside with a transaction and knows not to mess with the connection.   we have this recipe in our test suite so if it needs any adjustments they are covered.

I guess for the Session docs maybe the language should clarify "for an engine-bound Session". for a connection bound session there are some checks in place.

**Reply by mcclurem** [reply]
Wow, thanks for the instant response! Yeah the docs definitely confused me there but I'm now even realizing that you did in fact clarify in the docs for the unit test example at https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#tracking-transaction-state-with-events

Perhaps just add a see-also in the https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#session-level-vs-engine-level-transaction-control section?

Thanks again

**Comment by mcclurem** [comment]
Hey @zzzeek I /think/ I've found a bug with the example implementation but before I go and try to define the minimal test case I'm wondering if this is obvious:
I've implemented the recipe as shown and I'm making use of the end_savepoint function. What I've discovered is that during teardown I have to also remove the listener on "after_transaction_end" before calling session.close() or else the event fires and I get lots of errors. Perhaps this is an issue with the lifetime/scope of my pytest fixture (it's currently an auto-use fixture) but seems like the example should probably remove the listener before tearing down so that the initial session object is restored to original state?

**Reply by zzzeek** [reply]
there should be no need to remove the listener, calling session.close() should proceed normally.

**Comment by ezarowny** [comment]
How would one accomplish this if they weren't using Sessions and the ORM?

**Reply by CaselIT** [reply]
Connection at the moment doesn't support this, so calling commit/rollback on it will emit those actions in the driver connection

**Reply by ezarowny** [reply]
Thanks! This explains a lot of what I was experiencing and maybe a bit of my inability to work around it 😅
